cmd/joy/release.go

```diff
@@ -258,3 +266,111 @@ func NewReleaseRenderCmd() *cobra.Command {
 	cmd.Flags().BoolVar(&color, "color", term.IsTerminal(int(os.Stdout.Fd())), "toggle output with color")
 	return cmd
 }
+
+func NewGitQueryCommand(command string) *cobra.Command {
+	var (
+		source string
+		target string
+	)
+	cmd := &cobra.Command{
+		Use:  "git-" + command,
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cfg := config.FromContext(cmd.Context())
+
+			cat, err := catalog.Load(catalog.LoadOpts{Dir: cfg.CatalogDir})
+			if err != nil {
+				return err
+			}
+
+			release := args[0]
+
+			var sourceRelease, targetRelease *v1alpha1.Release
+
+			for _, cross := range cat.Releases.Items {
+				if cross.Name != release {
+					continue
+				}
+				for _, rel := range cross.Releases {
+					if rel == nil {
+						continue
+					}
+					switch rel.Environment.Name {
+					case source:
+						sourceRelease = rel
+					case target:
+						targetRelease = rel
+					}
+				}
+				break
+			}
+
+			if sourceRelease == nil {
+				return fmt.Errorf("no source found")
+			}
+			if targetRelease == nil {
+				return fmt.Errorf("no target found")
+			}
+
+			repoCache := filepath.Join(cfg.JoyCache, "src")
+			if err := os.MkdirAll(repoCache, 0o755); err != nil {
+				return fmt.Errorf("failed to ensure repo cache: %w", err)
+			}
+
+			var (
+				project = sourceRelease.Project
+				repo    = sourceRelease.Project.Spec.Repository
+			)
+
+			if _, err := os.Stat(filepath.Join(repoCache, project.Name)); err != nil {
+				if !errors.Is(err, os.ErrNotExist) {
+					return err
+				}
+
+				clone := exec.Command("gh", "repo", "clone", repo, project.Name)
+				clone.Stdout = os.Stdout
+				clone.Stderr = os.Stderr
+				clone.Dir = repoCache
+
+				if err := clone.Run(); err != nil {
+					return fmt.Errorf("failed to clone project: %w", err)
+				}
+			}
+
+			repoDir := filepath.Join(repoCache, project.Name)
+
+			getRevision := func(version string) string {
+				version = "v" + version
+				if branch := semver.Prerelease(version); branch != "" {
+					return branch
+				}
+				return path.Join(project.Spec.RepositorySubdirectory, version)
+			}
+
+			pull := exec.Command("git", "pull")
+			pull.Dir = repoDir
+
+			if err := pull.Run(); err != nil {
+				return fmt.Errorf("failed to pull project: %w", err)
+			}
+
+			expr := getRevision(sourceRelease.Spec.Version) + ".." + getRevision(targetRelease.Spec.Version)
+
+			gitCommand := exec.Command("git", command, expr)
```

Comment:
Can we make it that the user can pass additional flags? e.g.: `-w`?

Reply:
It would also be nice if we output the full command used (and the working dir) so the user can execute it themselves or re-use the commit range for example.

Reply:
to the prior, maybe! It being arbitrary flags might be hard with cobra. It's something I could do (and do do) with the standard flag package all the time, but yeah.

Reply:
found a way possibly